gax/src/main/java/com/google/api/gax/batching/v2/BatchingDescriptor.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.batching.v2;
+
+import com.google.api.core.SettableApiFuture;
+import java.util.List;
+
+/**
+ * An adapter that packs and unpacks the elements in and out of batch requests and responses.
+ *
+ * <p>This interface should be implemented by either a service specific client or autogenerated by
+ * gapic-generator.
+ *
+ * <p>Example implementation:
+ *
+ * <pre>{@code
+ * class ListBatchingDescriptor implements BatchingDescriptor<String, String, List<String>,
+ * List<String>> {
+ *
+ *   RequestBuilder<String, List<String>> newRequestBuilder(List<String> prototype) {
+ *     return new RequestBuilder<String, List<String>>() {
+ *       List<String> list = prototype.clone();
+ *
+ *       void add(String element) {
+ *         list.add(element);
+ *       }
+ *
+ *       List<String> build() {
+ *         return list.clone();
+ *       }
+ *     };
+ *   }
+ *
+ *   void splitResponse(List<String> batchResponse, List<SettableApiFuture<String>> batch) {
+ *     for (int i = 0; i < batchResponse.size(); i++) {
+ *       batch.get(i).set(batchResponse.get(i);
```

Comment:
Not all `Lists` are random-access lists (linked lists aren't), so this loop can be `O(n^2)`. What exactly is this sample trying to show? 
I agree that it is nitpicking but technically it should either ensure that the lists implement `RandomAccess` or do something like:
```java
    Iterator<String> respIt = batchResponse.iterator();
    Iterator<String> it = batchResponse.iterator();

    //should it also check that it.hasNext()? the original implementation does not.
    while(respIt.hasNext()) { 
      it.next().set(respIt.next());
    }
```

Reply:
I would really prefer to keep it as is. The point of the example is to help understand the semantics of this interface and to succinctly convey the relationship between parameters. Accounting for the performance of random access lists just muddies the whole thing up.